### PR TITLE
[BlockStore] Split BlockStoreLoop #206

### PR DESCRIPTION
**Pull request by fassadlr**
Split BlockStoreLoop into 4 logical areas and add unit tests using a successor pattern.

Add Steps

CheckExists
Reorganise
TryPending
DownloadBlocks

### CODE REVIEW

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreLoopStep.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                if (stepResult.ShouldBreak)
+                {
+                    result = stepResult;
+                    break;
+                }
+
+                if (stepResult.ShouldContinue)
+                {
+                    result = stepResult;
+                    break;
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreLoopStep.cs]: this should be a ```continue;``` no?

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreLoopStep.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            return this;
+        }
+
+        internal BlockStoreLoopStepResult Continue()
+        {
+            this.ShouldContinue = true;
+            return this;
+        }
+
+        internal BlockStoreLoopStepResult Next()
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreLoopStep.cs]: I think a static Next() would be more readable instead of a doing 
```new BlockStoreLoopStepResult().Next```
```  BlockStoreLoopStepResult.Next()```

Same goes for the other two methods

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        internal DownloadBlockStep(BlockStoreLoop blockStoreLoop)
+            : base(blockStoreLoop)
+        {
+        }
+
+        private int insertBlockSize = 0;
+        private int stallCount = 0;
+        private List<BlockPair> blockPairsToStore = new List<BlockPair>();
+
+        internal override async Task<BlockStoreLoopStepResult> Execute(ChainedBlock nextChainedBlock, CancellationToken cancellationToken, bool disposeMode)
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: My take here is preserving the original code, to me it was more readable and then you don't need to preserve state between methods. 
You then wont need a dedicated DownloadBlockStepResult 

But you might have a point with breaking the code further.

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+            this.BlockStoreLoop.BlockPuller.AskBlock(nextChainedBlock);
+
+            var canDownload = true;
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                if (canDownload)
+                {
+                    var askAndEnqueueResult = await AskAndEnqueueBlocksToDownload(chainedBlockStackToDownload, nextChainedBlock, canDownload);
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: I would call it FindBlocksToDownload

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    //      Collecting downloaded blocks and persisting them as a batch to the BlockRepository
+
+    internal sealed class DownloadBlockStep : BlockStoreLoopStep
+    {
+        internal DownloadBlockStep(BlockStoreLoop blockStoreLoop)
+            : base(blockStoreLoop)
+        {
+        }
+
+        private int insertBlockSize = 0;
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: This variables should go in a context so we dont carry state in the step

**Comment by dangershony** [comment]
I would suggest perhaps also adding more detailed comments

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs [hunk 6]
```diff
@@ -1,347 +1,187 @@
...
-		public ConcurrentDictionary<uint256, BlockPair> PendingStorage { get; }
-	    public ChainedBlock StoredBlock { get; private set; }
-
-        public BlockStoreLoop(ConcurrentChain chain, 
-            BlockRepository blockRepository, 
-            NodeSettings nodeArgs,
-			ChainState chainState,
-			StoreBlockPuller blockPuller,
+        public ConcurrentDictionary<uint256, BlockPair> PendingStorage { get; }
+        public ChainedBlock StoredBlock { get; internal set; }
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs]: Rename this to Tip or StoreTip
And maybe add a new method called .SetStoreTip(ChainedBlock )
It will also set the ChainState.HighestPersistedBlock

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        private int stallCount = 0;
+        private List<BlockPair> blockPairsToStore = new List<BlockPair>();
+
+        internal override async Task<BlockStoreLoopStepResult> Execute(ChainedBlock nextChainedBlock, CancellationToken cancellationToken, bool disposeMode)
+        {
+            if (disposeMode)
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: In this case we could even not call this step in dispose mode

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs [hunk 8]
```diff
@@ -1,347 +1,187 @@
...
             // When in IBD we may need many reads for the block key without fetching the block
             // So the repo starts with LazyLoadingOn = true, however when not anymore in IBD 
             // a read is normally done when a peer is asking for the entire block (not just the key) 
             // then if LazyLoadingOn = false the read will be faster on the entire block            
 
-            while (!token.IsCancellationRequested)
-			{
-				if (this.StoredBlock.Height >= this.ChainState.HighestValidatedPoW?.Height)
-					break;
+            var steps = new BlockStoreLoopStepChain(disposeMode);
```
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs]: Yes, the step chain will be created on every loop. I will move this to the Initialize() method.

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs [hunk 9]
```diff
@@ -1,347 +1,187 @@
...
 
-					var blockstoremove = new List<uint256>();
-					var remove = this.StoredBlock;
-					// reorg - we need to delete blocks, start walking back the chain
-					while (this.chain.GetBlock(remove.HashBlock) == null)
-					{
-						blockstoremove.Add(remove.HashBlock);
-						remove = remove.Previous;
-					}
+                BlockStoreLoopStepResult result = await steps.Execute(nextChainedBlock, cancellationToken);
```
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs]: Are you happy with how we are executing the steps now? I like iterating over the steps approach ;)

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreLoopStep.cs [hunk 10]
```diff
@@ -0,0 +1,87 @@
+﻿using NBitcoin;
+using Stratis.Bitcoin.Utilities;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Stratis.Bitcoin.Features.BlockStore.LoopSteps
+{
+    internal sealed class BlockStoreLoopStepChain
```
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreLoopStep.cs]: I will rename this.

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreLoopStep.cs [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        private List<BlockStoreLoopStep> steps = new List<BlockStoreLoopStep>();
+        private bool disposeMode;
```
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreLoopStep.cs]: I will create the disposeMode property as a argument to be passed into the Execute method()

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreLoopStep.cs (hunk 1 above)
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreLoopStep.cs]: The break is correct... cause the DownloadAndStoreBlocks() method will decide what to do with the result. I have refactored this little bit of code to be:

``if (stepResult.ShouldBreak || stepResult.ShouldContinue) 
{
                    result = stepResult;
                    `break;

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreLoopStep.cs (hunk 2 above)
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreLoopStep.cs]: Great minds think alike ;) I did have it as static methods initially. I will change it back.

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs (hunk 3 above)
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: OK I have reverted the original code back for this area so its more readable and remove the DownloadBlockStepResult

Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: But yes this class/logic will need to be split into 2 areas in the next PR.

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs (hunk 4 above)
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: This bit has been reverted as per previous comment.

File: Stratis.Bitcoin.Tests/BlockStore/LoopTests/BlockStoreLoopStepBaseTest.cs [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        private Mock<IConnectionManager> ConfigureConnectionManager()
+        {
+            var connectionManager = new Mock<IConnectionManager>();
+            connectionManager.Setup(c => c.ConnectedNodes).Returns(new NodesCollection());
+            connectionManager.Setup(c => c.NodeSettings).Returns(NodeSettings.Default());
+            connectionManager.Setup(c => c.Parameters).Returns(new NodeConnectionParameters());
+            return connectionManager;
+        }
+
+        internal void AppendBlock(ConcurrentChain chain, Block block)
```
Comment (fassadlr) [inline Stratis.Bitcoin.Tests/BlockStore/LoopTests/BlockStoreLoopStepBaseTest.cs]: I have changed this to:

```
        internal void AppendBlocks(ConcurrentChain chain, IEnumerable<Block> blocks)
        {
            foreach (var block in blocks)
            {
                if (chain.Tip != null)
                    block.Header.HashPrevBlock = chain.Tip.HashBlock;
                chain.SetTip(block.Header);
            }
        }
```

File: Stratis.Bitcoin.Tests/BlockStore/LoopTests/BlockStoreLoopStepBaseTest.cs [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+            chain.SetTip(block.Header);
+        }
+
+        internal List<Block> CreateBlocks(int amount)
+        {
+            var blocks = new List<Block>();
+            for (int i = 0; i < amount; i++)
+            {
+                Block block = CreateBlock(i);
+                block.Header.HashPrevBlock = blocks.Any() ? blocks.Last().GetHash() : Network.Main.GenesisHash;
```
Comment (fassadlr) [inline Stratis.Bitcoin.Tests/BlockStore/LoopTests/BlockStoreLoopStepBaseTest.cs]: Thanks! Thats a much better way yes :)

File: Stratis.Bitcoin.Tests/BlockStore/LoopTests/BlockStoreLoopStepBaseTest.cs [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+            var block = new Block();
+
+            for (int j = 0; j < 100; j++)
+            {
+                var trx = new Transaction();
+
+                block.AddTransaction(new Transaction());
+
+                trx.AddInput(new TxIn(Script.Empty));
+                trx.AddOutput(Money.COIN + j + blockNumber, new Script(Guid.NewGuid().ToByteArray()
```
Comment (fassadlr) [inline Stratis.Bitcoin.Tests/BlockStore/LoopTests/BlockStoreLoopStepBaseTest.cs]: Thanks! Thats a much shorter way yes :)

File: Stratis.Bitcoin.Tests/BlockStore/LoopTests/BlockStoreLoopStepReorganiseTest.cs [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+        public void CanExecute_Reorganise()
+        {
+            List<Block> blocks = CreateBlocks(15);
+
+            // The BlockRepository has 15 blocks stored
+            using (var blockRepository = new BlockRepository(Network.Main, TestBase.AssureEmptyDirAsDataFolder(@"BlockStore\LoopTest_Reorganise")))
+            {
+                blockRepository.PutAsync(blocks.Last().GetHash(), blocks).GetAwaiter().GetResult();
+
+                // The BlockRepository has 10 blocks appended
```
Comment (fassadlr) [inline Stratis.Bitcoin.Tests/BlockStore/LoopTests/BlockStoreLoopStepReorganiseTest.cs]: I have fixed this.

File: Stratis.Bitcoin.Tests/BlockStore/LoopTests/BlockStoreLoopStepBaseTest.cs [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+            INodeLifetime nodeLifeTime = new Mock<INodeLifetime>().Object;
+
+            var blockStoreLoop = new BlockStoreLoop(
+                asyncLoopFactory,
+                blockPuller,
+                blockRepository,
+                null,
+                chain,
+                chainState.Object,
+                NodeSettings.FromArguments(new string[] { }),
```
Comment (fassadlr) [inline Stratis.Bitcoin.Tests/BlockStore/LoopTests/BlockStoreLoopStepBaseTest.cs]: I have made this change.. I suspect that's why some of my tests were failed.

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs (hunk 7 above)
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: I cant refactor this right now as some of the logic in the steps depends on it. I.e not every step will NOT execute if disposeMode is true.

We can possible refactor this in the next PR?

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/ReorganiseBlockRepositoryStep.cs [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+                }
+
+                //Delete blocks from the repository
+                await this.BlockStoreLoop.BlockRepository.DeleteAsync(blockToDelete.HashBlock, blocksToDelete);
+
+                //Set the last stored block to the last found chained block
+                this.BlockStoreLoop.StoredBlock = blockToDelete;
+                this.BlockStoreLoop.ChainState.HighestPersistedBlock = this.BlockStoreLoop.StoredBlock;
+
+                blocksToDelete.Clear();
```
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/ReorganiseBlockRepositoryStep.cs]: We can possible refactor this in the next PR?

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs (hunk 6 above)
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs]: We can possible refactor this in the next PR?

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/ReorganiseBlockRepositoryStep.cs (hunk 17 above)
Comment (fassadlr) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/ReorganiseBlockRepositoryStep.cs]: I have remove this now :)

File: Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs (hunk 9 above)
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs]: yes and fine if you keep it in the step chain

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreLoopStep.cs (hunk 1 above)
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/BlockStoreLoopStep.cs]: Ah right yes and the caller will do the continue; fair enough

File: Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs (hunk 3 above)
Comment (dangershony) [inline Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs]: ok yeap fine with me, the logic can be indeed broken better in another PR

**Comment by dangershony** [comment]
Nice I assume the tests and integration tests passed?
